test/AdaptiveSpinLockTest.cpp

```diff
@@ -62,27 +62,94 @@ TEST(AdaptiveSpinLockTest, WorksWithStdUniqueLock) {
   lock.unlock();
 }
 
-const int thread_num = 7;
-int val;
-pipert::AdaptiveSpinLock* g_lock = nullptr;
-
-void SharedIncrementor(int times) {
-  ASSERT_TRUE(g_lock);
-  for (int i = 0; i < times; i++) {
-    std::lock_guard<pipert::AdaptiveSpinLock> guard(*g_lock);
-    val++;
+namespace {
+
+class SharedIncrementor {
+ public:
+  void StartThreads(void func (SharedIncrementor&)) {
+    for(unsigned int i = 0; i < threads_.size(); ++i)
+      threads_[i] = std::thread(func, std::ref(*this));
+  }
+  void JoinThreads() {
+    for(unsigned int i = 0; i < threads_.size(); ++i)
+      threads_[i].join();
+  }
+
+  void SetThreads(const int count) { threads_.resize(count); }
+  int GetThreads() const { return threads_.size(); }
+  void SetCycles(const int cycles) { cycles_ = cycles; }
+  int GetCycles() const { return cycles_; }
+  void ResetVal() { val_ = 0; }
+  void IncrementVal() { val_++; }
+  int GetVal() const { return val_; }
+
+ private:
+  std::vector<std::thread> threads_;
+  int cycles_;
+  int val_;
+};
+
+static SharedIncrementor& SharedIncrementorInstance() {
+  static SharedIncrementor si;
+  return si;
+}
+
+void LockEveryStep(SharedIncrementor& shinc) {
+  static pipert::AdaptiveSpinLock slock(2);
```

Comment:
This will probably lead to unknown consequences. The problem that you want one spinlock for the whole test. However here, a global one is created for all the tests.
But what is worse (and this can cause a crash) is that there is a check which decides whether this static instance should be created or not depending on its state. But I think, this is not guaranteed to run in a thread safe way. However, you start multiple threads and you expect this "was the static initialized" decision to be made in a synchronized way. As this code will not use synchronization, this will do a data race.

Reply:
Also true for the other similar functions.

Reply:
I think these should not cause any issues.

"However here, a global one is created for all the tests."
I don't think that would be the case. Block (function) scope static variables belonging to different functions will definitely get different instances initialized for them the first time control enters the function. They might be called the same, but I don't know why they wouldn't get different instances for each function. They use the same identifier, but exist in different scopes, and thus they must denote different entities.

"But I think, this is not guaranteed to run in a thread safe way."
I tried to research that, and static local variable initialization should be thread safe.
From https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables :

"If multiple threads attempt to initialize the same static local variable concurrently, the initialization occurs exactly once (similar behavior can be obtained for arbitrary functions with std::call_once)."

It seems to me that a decade ago protection of static local variables was added to gcc when it would not have been thread safe otherwise, but since C++11, the above quote should hold true. You are right in being suspicious, though, as this seems like something that should not be taken for granted.

Reply:
On the block function: Yes. I was referring to the fact that one single created instance in one single function will be used in different unit tests. So despite having them separated per function, they are not separate between tests.

On the thread-safe static init. This is bad for us because we are doing a synchronization in a test where we are testing synchronization itself. With this, we are forcing our threads to do other sync activities besides ours. But anyway, the idea would be to have clean tests: Everything used is created before the test and is shut down after the test. Nothing is reused between tests. This is also dangerous thinking on what could happen if we change the implementation to have more state inside.
But anyway, a clean start and shutdown is perfectly manageable with the fixtures you created.

Reply:
Ah, I see what you mean. It really is possible that if multiple tests use the same created instance (because the instance was statically initialized inside the function), testing ends up being unpragmatic. It really is better to recreate the spinlock for each test with build-up and tear-down per test case, so each test start with a completely clean slate. Now that the helper fixture is also created per test, it makes sense to put it back there to be constructed and destructed by it.

About thread safety: yes, it's kind of ironic to have a small point of synchronization between threads that enter the function for the first time, when testing synchronization. Even though it only happens when the static variable is uninitialized, and is reduced to a simple check afterwards, if I remember right.